finrl/env/env_stocktrading_cashpenalty.py

```diff
@@ -247,27 +275,40 @@ def step(self, actions):
             self.account_information["reward"].append(reward)
 
             # multiply action values by our scalar multiplier and save
-            actions = actions * self.hmax
-            self.actions_memory.append(actions)
-
-            # scale cash purchases to asset # changes
-            actions = actions / closings
+            actions = actions * self.hmax 
+            #buy/sell only if the price is > 0 (no missing data in this particular date 
+            actions = np.where(closings>0,actions,0)
+
+            if self.turbulence_threshold is not None:
+                # if turbulence goes over threshold, just clear out all positions
+                if self.turbulence>=self.turbulence_threshold:
+                    actions = -(np.array(holdings) * closings)
+                    self.log_step(reason="TURBULENCE")
+
+            # scale cash purchases to asset
+            if self.discrete_actions:
+                #convert into integer because we can't buy fraction of shares
+                actions = np.where(closings>0,actions//closings,0)
+                actions = (actions.astype(int))
+            else:
+                actions = np.where(closings>0,actions/closings,0)
             
-
             # clip actions so we can't sell more assets than we hold
             actions = np.maximum(actions, -np.array(holdings))
+
             self.transaction_memory.append(actions)
+            self.actions_memory.append(actions*closings)
```

Comment:
Can we move this up so it's not subject to any of our modifications? Actions_memory is supposed to capture what the model is trying to do, where transaction memory indicates what it could do.

Reply:
Ahh I see, okay sure